### DISCUSSION TITLE
Tangoboard wiring

### DISCUSSION THREAD

**Discussion by kzali** [multigcs/LinuxCNC-RIO#24] [Q&A]
Hi Oliver a few, perhaps silly questions!

1. Just noticed in the connectors.md You have suggested 5 pin connectors, while the Tango board has 4 pin connectors for the Joints. What would be a modification in wiring?  (Most external drivers have 6 wire inputs perhaps with 2 control inputs being ground.)

2. What is the jumpers, for, in the image below? Is it power select?
![image](https://github.com/multigcs/LinuxCNC-RIO/assets/7270401/7ecc2882-c7d1-410a-a0f0-8e2d873d44ae)

3. Is there a sample wiring image for say a 3 axis mill with maybe ,TB6600 or similar and how is the tangnano9k board  powered?

Many thanks for your help.

**Comment by multigcs** [comment]
1. in the future, i will add a GND pin to the stepper-outputs, because i have 'mks servo42c' drivers that need 5V logic power and i like to have all the needed pins on one connector.
2. With this two jumper , you can switch the voltage on the Input-Connectors between 5V and 12-24V power
3. same like here: https://i1.wp.com/www.rainydaymagazine.com/RDM2012/RainyDayProjects/BlueChick/BOBSetup/WiringBig.jpg

**Comment by kzali** [comment]
> 1. in the future, i will add a GND pin to the stepper-outputs, because i have 'mks servo42c' drivers that need 5V logic power and i like to have all the needed pins on one connector.
> 2. With this two jumper , you can switch the voltage on the Input-Connectors between 5V and 12-24V power
> 3. same like here: https://i1.wp.com/www.rainydaymagazine.com/RDM2012/RainyDayProjects/BlueChick/BOBSetup/WiringBig.jpg

Many thanks that makes sense.  I could see the jumpers for your external drivers on your tango board short video.  I will diagram my setup.  Still don't understand how the Tangnano9k is powered and the purpose of the jumpers on the board.  Maybe I will look at the schematic and see how it works.  Also my test TB6600 drivers take 5v. control signals.

**Comment by multigcs** [comment]
to power the fpga, you can use the 5V pins next to the spi-connector, or for directly via the USB port on the TangNano9K.

**Comment by kzali** [comment]
Hello Oliver
Just did a git pull for updates but cot the following error:
kam@kzbuntu:~/LinuxCNC-RIO$ git pull
remote: Enumerating objects: 198, done.
remote: Counting objects: 100% (198/198), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 141 (delta 100), reused 132 (delta 92), pack-reused 0
Receiving objects: 100% (141/141), 28.31 KiB | 3.54 MiB/s, done.
Resolving deltas: 100% (100/100), completed with 53 local objects.
From https://github.com/multigcs/LinuxCNC-RIO
   04ef0be..ca33312  main       -> origin/main
Updating 04ef0be..ca33312
Fast-forward
 .gitignore                                         |   2 +
 Makefile                                           |   7 +-
 buildtool.py                                       | 264 +++---------
 generators/firmware/buildsys.py                    |   3 +
 generators/firmware/pins.py                        |   4 +-
 .../linuxcnc_component/linuxcnc_component.py       |   4 +-
 generators/linuxcnc_component/rio.c                |  53 ++-
 generators/qtgui/qtgui.py                          |   5 +-
 plugins/din_bit/plugin.py                          |   4 +-
 plugins/din_toggle/plugin.py                       |   4 +-
 plugins/dout_bit/plugin.py                         |   4 +-
 plugins/expansion_shiftreg/plugin.py               | 107 +++--
 plugins/interface_spislave/plugin.py               |  25 +-
 plugins/interface_uart/README.md                   |  17 +
 plugins/interface_uart/interface_uart.v            | 230 ++++++++++
 plugins/interface_uart/plugin.py                   |  66 +++
 plugins/joint_pwmdir/plugin.py                     |   2 +-
 plugins/joint_rcservo/plugin.py                    |  10 +-
 plugins/joint_stepper/plugin.py                    |  14 +-
 plugins/vfdbridge/plugin.py                        |  11 +-
 plugins/vin_ads1115/plugin.py                      |  12 +-
 plugins/vin_frequency/plugin.py                    |   4 +-
 plugins/vin_lm75/plugin.py                         |   5 +-
 plugins/vin_pulsecounter/plugin.py                 |  11 +-
 plugins/vin_pwmcounter/plugin.py                   |   4 +-
 plugins/vin_quadencoder/plugin.py                  |  24 +-
 plugins/vin_quadencoderz/plugin.py                 |  20 +-
 plugins/vin_sonar/plugin.py                        |  19 +-
 plugins/vout_frequency/plugin.py                   |   6 +-
 plugins/vout_pwm/plugin.py                         |  11 +-
 plugins/vout_rcservo/plugin.py                     |   7 +-
 plugins/vout_sinepwm/plugin.py                     |  12 +-
 plugins/vout_spipoti/plugin.py                     |  22 +-
 plugins/vout_udpoti/plugin.py                      |  18 +-
 projectLoader.py                                   | 148 +++++++
 qt-testgui.py                                      | 474 +++++++++++++++++++++
 qtsetup.py                                         |   2 +-
 tests/data/tangnano9k_1/config.json                | 306 +++++++++++++
 tests/data/tangnano9k_1/rio.h                      | 171 ++++++++
 tests/data/tangnano9k_1/rio.hal                    |  81 ++++
 tests/data/tangnano9k_1/rio.ini                    | 219 ++++++++++
 tests/data/tangnano9k_1/rio.v                      | 271 ++++++++++++
 tests/test_generator.py                            |  24 ++
 43 files changed, 2324 insertions(+), 383 deletions(-)
 create mode 100644 plugins/interface_uart/README.md
 create mode 100644 plugins/interface_uart/interface_uart.v
 create mode 100644 plugins/interface_uart/plugin.py
 create mode 100644 projectLoader.py
 create mode 100644 qt-testgui.py
 create mode 100644 tests/data/tangnano9k_1/config.json
 create mode 100644 tests/data/tangnano9k_1/rio.h
 create mode 100644 tests/data/tangnano9k_1/rio.hal
 create mode 100644 tests/data/tangnano9k_1/rio.ini
 create mode 100644 tests/data/tangnano9k_1/rio.v
 create mode 100644 tests/test_generator.py
kam@kzbuntu:~/LinuxCNC-RIO$ bash patching-oss-cad-suite.sh /home/kam/oss-cad-suite/
ERROR: Apycula-0.8.2a2.dev1+g8e31050-py3.8.egg not found in /home/kam/oss-cad-suite//lib/python3.8/site-packages/
please user the right version of oss-cad-suite: 2023-07-20
kam@kzbuntu:~/LinuxCNC-RIO$ cp configs/Tangoboard/config.json configs/Tangoboard/config-mycnc.json
kam@kzbuntu:~/LinuxCNC-RIO$ python3 buildtool.py configs/Tangoboard/config-mycnc.json
%Error: plugins/interface_uart/interface_uart.v:1:1: Unknown verilator lint message code: 'WIDTHTRUNC', in '/*verilator lint_off WIDTHTRUNC*/'
    1 | /*verilator lint_off WIDTHTRUNC*/ 
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
%Error: plugins/interface_uart/interface_uart.v:2:1: Unknown verilator lint message code: 'WIDTHEXPAND', in '/*verilator lint_off WIDTHEXPAND*/'
    2 | /*verilator lint_off WIDTHEXPAND*/ 
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
%Error: Exiting due to 2 error(s)
generating files in Output/Tangoboard
generating linux-cnc config: axis
Traceback (most recent call last):
  File "/home/kam/LinuxCNC-RIO/buildtool.py", line 74, in <module>
    main(args.configfile, args.outputdir)
  File "/home/kam/LinuxCNC-RIO/buildtool.py", line 60, in main
    generator.generate(project)
  File "/home/kam/LinuxCNC-RIO/generators/linuxcnc_config/linuxcnc_config.py", line 1646, in generate
    generate_rio_gui(project)
  File "/home/kam/LinuxCNC-RIO/generators/linuxcnc_config/linuxcnc_config.py", line 1454, in generate_rio_gui
    elif vin_net == "spindle.0.speed-out" and gui == "qtdragon":
UnboundLocalError: local variable 'vin_net' referenced before assignment

**Reply by multigcs** [reply]
upps, sorry, i fixed it

**Reply by kzali** [reply]
Thanks will do a git pull and try.  Also doing a write up and wiring for the tango board.  as I go along.

**Comment by kzali** [comment]
Okay so tested it again and this time the Output folder was fully populated but did get the following errors:

kam@kzbuntu:~/LinuxCNC-RIO$ git pull
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 5 (delta 4), reused 5 (delta 4), pack-reused 0
Unpacking objects: 100% (5/5), 446 bytes | 63.00 KiB/s, done.
From https://github.com/multigcs/LinuxCNC-RIO
   ca33312..2aa7ede  main       -> origin/main
Updating ca33312..2aa7ede
Fast-forward
 generators/linuxcnc_config/linuxcnc_config.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
kam@kzbuntu:~/LinuxCNC-RIO$ cp configs/Tangoboard/config.json configs/Tangoboard/config-mycnc.json
kam@kzbuntu:~/LinuxCNC-RIO$ python3 buildtool.py configs/Tangoboard/config-mycnc.json
%Error: plugins/interface_uart/interface_uart.v:1:1: Unknown verilator lint message code: 'WIDTHTRUNC', in '/*verilator lint_off WIDTHTRUNC*/'
    1 | /*verilator lint_off WIDTHTRUNC*/ 
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
%Error: plugins/interface_uart/interface_uart.v:2:1: Unknown verilator lint message code: 'WIDTHEXPAND', in '/*verilator lint_off WIDTHEXPAND*/'
    2 | /*verilator lint_off WIDTHEXPAND*/ 
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
%Error: Exiting due to 2 error(s)
generating files in Output/Tangoboard
generating linux-cnc config: axis
generating linux-cnc component
generating qtgui
generating firmware

**Reply by multigcs** [reply]
fixed, but you can normally ignore verilator errors

btw:  your 'verilator' version looks a little bit too old: 'Unknown verilator lint message code'

maybe you have installed 'verilator' via package-system (apt), then you can remove it:
```
sudo apt-get remove verilator
```
the oss-cad-suite have it's own in a newer version

**Reply by kzali** [reply]
I never installed Verilator, I just installed litex

I guess I have to update that.  This is all very new to me.  Thanks for your help

**Comment by kzali** [comment]
![tangoboard-spi](https://github.com/multigcs/LinuxCNC-RIO/assets/7270401/8bbd5f87-b93f-4c64-9316-59d54543d1c0)
hi
on the Tango board which pin is CS and which is SCLK
![eth01](https://github.com/multigcs/LinuxCNC-RIO/assets/7270401/91ac7787-d294-420f-9fc0-a024192bae90)

**Reply by multigcs** [reply]
SCLK -> CLK (SerialClock -> Clock)
CS -> SEL (ChipSelect -> Select)

sorry  for the different names :(

**Comment by kzali** [comment]
Thank you
